# EEG Preprocessing

## Goal

Prepare raw EEG recordings for motor imagery classification by extracting event markers, filtering the signal, and creating epochs.

In [2]:
import mne
import matplotlib.pyplot as plt

raw = mne.io.read_raw_edf("../data/S001R06.edf", preload=True)

print(raw)

Extracting EDF parameters from ../data/S001R06.edf...
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19999  =      0.000 ...   124.994 secs...
<RawEDF | S001R06.edf, 64 x 20000 (125.0 s), ~9.8 MiB, data loaded>


In [3]:
events, event_id = mne.events_from_annotations(raw)

print(event_id)
print(events[:10])

Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
{np.str_('T0'): 1, np.str_('T1'): 2, np.str_('T2'): 3}
[[   0    0    1]
 [ 672    0    3]
 [1328    0    1]
 [2000    0    2]
 [2656    0    1]
 [3328    0    2]
 [3984    0    1]
 [4656    0    3]
 [5312    0    1]
 [5984    0    2]]


In [4]:
event_id_clean = {
    "class_1": 2,
    "class_2": 3
}

event_id_clean

{'class_1': 2, 'class_2': 3}

In [5]:
raw_filtered = raw.copy().filter(8, 30)

Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 8 - 30 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 8.00
- Lower transition bandwidth: 2.00 Hz (-6 dB cutoff frequency: 7.00 Hz)
- Upper passband edge: 30.00 Hz
- Upper transition bandwidth: 7.50 Hz (-6 dB cutoff frequency: 33.75 Hz)
- Filter length: 265 samples (1.656 s)



In [6]:
epochs = mne.Epochs(
    raw_filtered,
    events,
    event_id=event_id_clean,
    tmin=0,
    tmax=4,
    baseline=None,
    preload=True
)

print(epochs)

Not setting metadata
15 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 15 events and 641 original time points ...
0 bad epochs dropped
<Epochs | 15 events (all good), 0 – 4 s (baseline off), ~4.7 MiB, data loaded,
 'class_1': 7
 'class_2': 8>


In [8]:
print(epochs)
print(epochs.event_id)
print(epochs.get_data().shape)

<Epochs | 15 events (all good), 0 – 4 s (baseline off), ~4.7 MiB, data loaded,
 'class_1': 7
 'class_2': 8>
{'class_1': 2, 'class_2': 3}
(15, 64, 641)


In [9]:
import mne

files = [
    "../data/S001R06.edf",
    "../data/S001R10.edf",
    "../data/S001R14.edf",
]

raws = [mne.io.read_raw_edf(f, preload=True) for f in files]
raw = mne.concatenate_raws(raws)

print(raw)

Extracting EDF parameters from ../data/S001R06.edf...
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19999  =      0.000 ...   124.994 secs...
Extracting EDF parameters from ../data/S001R10.edf...
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19999  =      0.000 ...   124.994 secs...
Extracting EDF parameters from ../data/S001R14.edf...
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19999  =      0.000 ...   124.994 secs...
<RawEDF | S001R06.edf, 64 x 60000 (375.0 s), ~29.3 MiB, data loaded>


In [10]:
events, event_id = mne.events_from_annotations(raw)

print(event_id)
print(events.shape)
print(events[:10])

Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
{np.str_('T0'): 1, np.str_('T1'): 2, np.str_('T2'): 3}
(90, 3)
[[   0    0    1]
 [ 672    0    3]
 [1328    0    1]
 [2000    0    2]
 [2656    0    1]
 [3328    0    2]
 [3984    0    1]
 [4656    0    3]
 [5312    0    1]
 [5984    0    2]]


In [11]:
raw_filtered = raw.copy().filter(8, 30)

Filtering raw data in 3 contiguous segments
Setting up band-pass filter from 8 - 30 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 8.00
- Lower transition bandwidth: 2.00 Hz (-6 dB cutoff frequency: 7.00 Hz)
- Upper passband edge: 30.00 Hz
- Upper transition bandwidth: 7.50 Hz (-6 dB cutoff frequency: 33.75 Hz)
- Filter length: 265 samples (1.656 s)



In [12]:
event_id_clean = {
    "class_1": 2,
    "class_2": 3
}

epochs = mne.Epochs(
    raw_filtered,
    events,
    event_id=event_id_clean,
    tmin=0,
    tmax=4,
    baseline=None,
    preload=True
)

print(epochs)
print(epochs.get_data().shape)

Not setting metadata
45 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 45 events and 641 original time points ...
0 bad epochs dropped
<Epochs | 45 events (all good), 0 – 4 s (baseline off), ~14.1 MiB, data loaded,
 'class_1': 21
 'class_2': 24>
(45, 64, 641)


In [ ]:
epochs.save("../data/subject1_motor_imagery-epo.fif", overwrite=True)

/var/folders/f5/nrt9d59s6pjbdtg5wz2y15v00000gn/T/ipykernel_79885/2439700775.py:1: RuntimeWarning: This filename (../data/subject1_motor_imagery_epochs.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  epochs.save("../data/subject1_motor_imagery_epochs.fif", overwrite=True)


[PosixPath('/Users/ceciliajia/Desktop/eeg-motor-imagery-classification/notebooks/../data/subject1_motor_imagery_epochs.fif')]

## Preprocessing Result

After combining runs 6, 10, and 14 for Subject 1, I extracted 45 motor imagery epochs. Each epoch contains 64 EEG channels and 641 time points. The data was filtered between 8 and 30 Hz, which has the mu and beta rhythms which are most commonly used in motor imagery BCI classification.